ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/Snapshots.hs

```diff
@@ -180,32 +194,98 @@ diskSnapshotIsPermanent = isJust . dsSuffix
-      ExceptT
-    . readIncremental hasFS decoder
-    . snapshotToPath
+  -> ExceptT ReadSnapshotErr m (ExtLedgerState blk)
+readSnapshot someHasFS decLedger decHash snapshotName = do
+  !snapshotCRC <- readCRC someHasFS (snapshotToChecksumPath snapshotName)
+  ExceptT $
+    readIncremental someHasFS decoder (snapshotToPath snapshotName) >>= \case
+       Left e  -> pure $ Left (ReadSnaphotFailed e)
+       Right (ledgerState, checksumAsRead) ->
+           if checksumAsRead /= snapshotCRC
+           then pure $ Left ReadSnaphotDataCorruption
+           else pure (Right ledgerState)
   where
     decoder :: Decoder s (ExtLedgerState blk)
     decoder = decodeSnapshotBackwardsCompatible (Proxy @blk) decLedger decHash
 
--- | Write snapshot to disk
+    readCRC ::
+      SomeHasFS m
+      -> FsPath
+      -> ExceptT ReadSnapshotErr m CRC
+    readCRC (SomeHasFS hasFS) crcPath = ExceptT $ do
+        crcExists <- doesFileExist hasFS crcPath
+        if not crcExists
+          then pure (Left $ ReadSnapshotNoChecksumFile crcPath)
+          else do
+            withFile hasFS crcPath ReadMode $ \h -> do
+              str <- BSL.toStrict <$> hGetAll hasFS h
+              if not (BSC.length str == 8 && BSC.all isHexDigit str)
+                then pure (Left $ ReadSnapshotInvalidChecksumFile crcPath)
+                else pure . Right . CRC $ fromIntegral (hexdigitsToInt str)
+        -- TODO: remove the functions in the where clause when we start depending on lsm-tree
+      where
+        isHexDigit :: Char -> Bool
+        isHexDigit c = (c >= '0' && c <= '9')
+                    || (c >= 'a' && c <= 'f') --lower case only
+
+        -- Precondition: BSC.all isHexDigit
+        hexdigitsToInt :: BSC.ByteString -> Word
+        hexdigitsToInt =
+            BSC.foldl' accumdigit 0
+          where
+            accumdigit :: Word -> Char -> Word
+            accumdigit !a !c =
+              (a `shiftL` 4) .|. hexdigitToWord c
+
+
+        -- Precondition: isHexDigit
+        hexdigitToWord :: Char -> Word
+        hexdigitToWord c
+          | let !dec = fromIntegral (ord c - ord '0')
+          , dec <= 9  = dec
+
+          | let !hex = fromIntegral (ord c - ord 'a' + 10)
+          , otherwise = hex
+
+-- | Write a ledger state snapshot to disk
+--
+--   This function writes two files:
+--   * the snapshot file itself, with the name generated by @'snapshotToPath'@
+--   * the checksum file, with the name generated by @'snapshotToChecksumPath'@
 writeSnapshot ::
      forall m blk. MonadThrow m
   => SomeHasFS m
   -> (ExtLedgerState blk -> Encoding)
   -> DiskSnapshot
   -> ExtLedgerState blk -> m ()
 writeSnapshot (SomeHasFS hasFS) encLedger ss cs = do
-    withFile hasFS (snapshotToPath ss) (WriteMode MustBeNew) $ \h ->
-      void $ hPut hasFS h $ CBOR.toBuilder (encode cs)
+    !crc <- withFile hasFS (snapshotToPath ss) (WriteMode MustBeNew) $ \h -> do
+      (_, !crc) <- hPutAllCRC hasFS h $ CBOR.toLazyByteString (encode cs)
```

Comment:
One of these two bangs should be redundant I think. Perhaps looking at the Core output can show whether that is the case.

Reply:
Also I think it doesn't matter that much because the next line will write a file with the contents of the CRC so it will be evaluated soon.

Reply:
Agreed, both are redundant.